File: frontend/src/api/applications.ts
import { requestJson } from "./client";
import type {
  ApplicationCreatePayload,
  ApplicationFilters,
  ApplicationRecord,
  ApplicationReflectionPayload,
  ApplicationStats,
  ApplicationStatusHistoryRecord,
  ApplicationUpdatePayload,
  ListResponse,
} from "../types/api";

export function createApplication(
  payload: ApplicationCreatePayload,
): Promise<ApplicationRecord> {
  return requestJson<ApplicationRecord>("/api/applications", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });
}

export function listApplications(
  filters: ApplicationFilters = {},
): Promise<ListResponse<ApplicationRecord>> {
  const params = new URLSearchParams();
  if (filters.status) {
    params.set("status", filters.status);
  }
  if (filters.company) {
    params.set("company", filters.company);
  }
  if (filters.roleCategory) {
    params.set("role_category", filters.roleCategory);
  }
  if (filters.resumeVersionId) {
    params.set("resume_version_id", filters.resumeVersionId);
  }
  if (filters.jdId) {
    params.set("jd_id", filters.jdId);
  }
  if (filters.matchReportId) {
    params.set("match_report_id", filters.matchReportId);
  }
  if (filters.agentRunId) {
    params.set("agent_run_id", filters.agentRunId);
  }
  if (filters.priority) {
    params.set("priority", filters.priority);
  }
  if (filters.applyDateFrom) {
    params.set("apply_date_from", filters.applyDateFrom);
  }
  if (filters.applyDateTo) {
    params.set("apply_date_to", filters.applyDateTo);
  }
  if (filters.nextStepDateFrom) {
    params.set("next_step_date_from", filters.nextStepDateFrom);
  }
  if (filters.nextStepDateTo) {
    params.set("next_step_date_to", filters.nextStepDateTo);
  }
  const query = params.toString();
  return requestJson<ListResponse<ApplicationRecord>>(
    query ? `/api/applications?${query}` : "/api/applications",
  );
}

export function getApplication(
  applicationId: string,
): Promise<ApplicationRecord> {
  return requestJson<ApplicationRecord>(`/api/applications/${applicationId}`);
}

export function updateApplication(
  applicationId: string,
  payload: ApplicationUpdatePayload,
): Promise<ApplicationRecord> {
  return requestJson<ApplicationRecord>(`/api/applications/${applicationId}`, {
    method: "PATCH",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });
}

export function updateApplicationReflection(
  applicationId: string,
  payload: ApplicationReflectionPayload,
): Promise<ApplicationRecord> {
  return requestJson<ApplicationRecord>(
    `/api/applications/${applicationId}/reflection`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(payload),
    },
  );
}

export function listApplicationStatusHistory(
  applicationId: string,
): Promise<ListResponse<ApplicationStatusHistoryRecord>> {
  return requestJson<ListResponse<ApplicationStatusHistoryRecord>>(
    `/api/applications/${applicationId}/status-history`,
  );
}

export function getApplicationStats(): Promise<ApplicationStats> {
  return requestJson<ApplicationStats>("/api/applications/stats");
}

File: backend/app/api/applications.py
from fastapi import APIRouter, Depends, Query, Request, status
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.schemas.applications import (
    ApplicationCreateRequest,
    ApplicationRecord,
    ApplicationStats,
    ApplicationUpdateRequest,
)
from app.schemas.common import ApiResponse, ListResponse
from app.services import application_service


router = APIRouter(prefix="/api/applications", tags=["applications"])


@router.post(
    "",
    response_model=ApiResponse[ApplicationRecord],
    status_code=status.HTTP_201_CREATED,
)
async def create_application(
    request: Request,
    payload: ApplicationCreateRequest,
    db: Session = Depends(get_db),
) -> dict[str, object]:
    application = application_service.create_application(db, payload)
    return {"data": application, "request_id": request.state.request_id}


@router.get("", response_model=ApiResponse[ListResponse[ApplicationRecord]])
async def list_applications(
    request: Request,
    status: str | None = Query(default=None),
    company: str | None = Query(default=None),
    role_category: str | None = Query(default=None),
    resume_version_id: str | None = Query(default=None),
    jd_id: str | None = Query(default=None),
    agent_run_id: str | None = Query(default=None),
    db: Session = Depends(get_db),
) -> dict[str, object]:
    items = application_service.list_applications(
        db,
        status=status,
        company=company,
        role_category=role_category,
        resume_version_id=resume_version_id,
        jd_id=jd_id,
        agent_run_id=agent_run_id,
    )
    return {
        "data": ListResponse(items=items, total=len(items)),
        "request_id": request.state.request_id,
    }


@router.get("/stats", response_model=ApiResponse[ApplicationStats])
async def get_application_stats(
    request: Request,
    db: Session = Depends(get_db),
) -> dict[str, object]:
    stats = application_service.get_application_stats(db)
    return {"data": stats, "request_id": request.state.request_id}


@router.get("/{application_id}", response_model=ApiResponse[ApplicationRecord])
async def get_application(
    request: Request,
    application_id: str,
    db: Session = Depends(get_db),
) -> dict[str, object]:
    application = application_service.get_application(db, application_id)
    return {"data": application, "request_id": request.state.request_id}


@router.patch("/{application_id}", response_model=ApiResponse[ApplicationRecord])
async def update_application(
    request: Request,
    application_id: str,
    payload: ApplicationUpdateRequest,
    db: Session = Depends(get_db),
) -> dict[str, object]:
    application = application_service.update_application(db, application_id, payload)
    return {"data": application, "request_id": request.state.request_id}
